ufl/cell.py

```diff
@@ -1,298 +1,456 @@
+            self._sub_entity_types = [["vertex" for i in range(6)], ["interval" for i in range(9)], ["triangle", "quadrilateral", "quadrilateral", "quadrilateral", "triangle"], ["prism"]]
+        elif cellname == "pyramid":
+            self._num_cell_entities = (5, 8, 5, 1)
+            self._sub_entity_types = [["vertex" for i in range(5)], ["interval" for i in range(8)], ["quadrilateral", "triangle", "triangle", "triangle", "triangle"], ["pyramid"]]
+        elif cellname == "hexahedron":
+            self._num_cell_entities = (8, 12, 6, 1)
+            self._sub_entity_types = [["vertex" for i in range(8)], ["interval" for i in range(12)], ["quadrilateral" for i in range(6)], ["hexahedron"]]
+        else:
+            raise ValueError(f"Unsupported cell type: {cellname}")
 
-@attach_operators_from_hash_data
-class TensorProductCell(AbstractCell):
-    __slots__ = ("_cells",)
+        self._cellname = cellname
+        self._tdim = len(self._num_cell_entities) - 1
+        self._gdim = self._tdim if geometric_dimension is None else geometric_dimension
 
-    def __init__(self, *cells, **kwargs):
-        keywords = list(kwargs.keys())
-        if keywords and keywords != ["geometric_dimension"]:
-            raise ValueError(
-                "TensorProductCell got an unexpected keyword argument '%s'" %
-                keywords[0])
+        if not isinstance(self._gdim, numbers.Integral):
+            raise ValueError("Expecting integer geometric_dimension.")
+        if not isinstance(self._tdim, numbers.Integral):
+            raise ValueError("Expecting integer topological_dimension.")
+        if self._tdim > self._gdim:
+            raise ValueError("Topological dimension cannot be larger than geometric dimension.")
 
-        self._cells = tuple(as_cell(cell) for cell in cells)
+    def topological_dimension(self) -> int:
+        """Return the dimension of the topology of this cell."""
+        return self._tdim
+
+    def geometric_dimension(self) -> int:
+        """Return the dimension of the geometry of this cell."""
+        return self._gdim
+
+    def is_simplex(self) -> bool:
+        """Return True if this is a simplex cell."""
+        print(self._cellname)
+        return self._cellname in ["vertex", "interval", "triangle", "tetrahedron"]
+
+    def has_simplex_facets(self) -> bool:
+        """Return True if all the facets of this cell are simplex cells."""
+        return self._cellname in ["interval", "triangle", "quadrilateral", "tetrahedron"]
+
+    def num_sub_entities(self, dim: int) -> int:
+        """Get the number of sub-entities of the given dimension."""
+        try:
+            return self._num_cell_entities[dim]
+        except IndexError:
+            return 0
+
+    def sub_entities(self, dim: int) -> typing.List[AbstractCell]:
+        """Get the sub-entities of the given dimension."""
+        try:
+            return [Cell(t, self._gdim) for t in self._sub_entity_types[dim]]
```

Comment:
Could you just make these up front in the constructor rather than rebuilding every time one asks?